swag/restswag.py

```diff
@@ -0,0 +1,141 @@
+import inspect
+
+def decorated_controllers():
+    return []
+
+def collect_controllers(RootController):
+    """
+    Traverses project starting with RootController
+    and collects all RestControllers returning a tuple with their respective paths.
+    Concatenates decorated controllers and their paths to the list.
+    :return [(controllers, path),...]
+    """
+    stack = [(RootController, "/")]
+    queue = []
+    while stack:
+        restcontroller = stack.pop()
+        controllers = []
+        for ctrl in dir(restcontroller[0]):
+            if issubclass(type(getattr(restcontroller[0], ctrl)), rest.RestController):
+                controllers.append((getattr(restcontroller[0], ctrl), ctrl))
+        for controller in controllers:
+            sep = '/'
+            if restcontroller[1][-1] == sep: sep = ''
+            stack.append((controller[0], restcontroller[1] + sep + controller[1]))
+        queue.append(restcontroller)
+
+    queue + decorated_controllers()
+    return queue
+
+def collect_methods(controllers):
+    """
+    collects all restful method paths from all restcontrollers in controllers
+    and returns a dictionary with their method-names and methods.
+    :return: {path: method-name: method}
+    """
+    methods = {}
+    for controller in controllers:
+
+        # going through standard pecan methods
+        if getattr(controller[0], 'get_one', None):
+            methodtuple = (controller[1] + "/{" +
+                               getattr(controller[0], 'get_one')._pecan['argspec'][0][1] + "}",
+                           ('get', getattr(controller[0], 'get_one')))
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'get_all', None):
+            methodtuple = (controller[1], ('get', getattr(controller[0], 'get_all')))
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'get', None):
+            methodtuple = (controller[1], ('get', getattr(controller[0], 'get')))
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'new', None):
+            pass
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'edit', None):
+            pass
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'post', None):
+            methodtuple = (controller[1], ('post', getattr(controller[0], 'post')))
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'put', None):
+            methodtuple = (controller[1], ('put', getattr(controller[0], 'put')))
+            methods = add_method(methods, methodtuple)
+        if getattr(controller[0], 'get_delete', None):
+            methodtuple = (controller[1] +  "/{" +
+                               getattr(controller[0], 'get_delete')._pecan['argspec'][0][1] + "}",
+                           ('delete', getattr(controller[0], 'get_delete')))
```

Comment:
same as the get_one entry, this could be refactored using the format command for the strings and removing the extra getattr calls